File: src/main/scala/fr/exemple/formation/dataframe/dataset/StrongTypingAndDatasetSecurity.scala
package fr.exemple.formation.dataframe.dataset

import org.apache.spark.sql.types.{DataTypes, StructField, StructType}
import org.apache.spark.sql.{Dataset, SparkSession}

object StrongTypingAndDatasetSecurity {

  def main(args: Array[String]): Unit = {
    // Créer une SparkSession
    val spark = SparkSession
      .builder
      .appName("StrongTypingAndDatasetSecurity")
      .master("local[*]")
      .getOrCreate()

    import spark.implicits._


    val schema = new StructType(Array[StructField](
      StructField("clientId", DataTypes.StringType),
      StructField("name", DataTypes.StringType),
      StructField("age", DataTypes.IntegerType)
    )
    )

    val clientsDS: Dataset[Client] = spark
      .read
      .option("header", "true")
      .schema(schema)
      .csv("src/main/resources/java/dataset/client.csv")
      .as[Client]

    // Afficher les clients
    clientsDS.show()

    val adultsDS = clientsDS.filter(x => x.isLegalAge)

    // Afficher les clients adultes
    adultsDS.show()

    // Fermer la SparkSession
    spark.stop()
  }

}

File: src/main/java/fr/example/formation/dataframe/dataset/StrongTypingAndDatasetSecurity.java
package fr.example.formation.dataframe.dataset;

import org.apache.spark.api.java.function.FilterFunction;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Encoders;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.sql.types.DataTypes;
import org.apache.spark.sql.types.Metadata;
import org.apache.spark.sql.types.StructField;
import org.apache.spark.sql.types.StructType;

public class StrongTypingAndDatasetSecurity {
    public static void main(String[] args) {
        // Créer une SparkSession
        SparkSession spark = SparkSession.builder()
                .appName("StrongTypingAndDatasetSecurity")
                .master("local[*]") // Mode local pour les tests
                .getOrCreate();


        StructType schema = new StructType(new StructField[]{
                new StructField("clientId", DataTypes.StringType, true, Metadata.empty()),
                new StructField("name", DataTypes.StringType, true, Metadata.empty()),
                new StructField("age", DataTypes.IntegerType, true, Metadata.empty()),

        });

        Dataset<Client> clientsDS =
                spark
                        .read()
                        .option("header", "true")
                        .schema(schema)
                        .csv("src/main/resources/java/dataset/client.csv")
                        .as(Encoders.bean(Client.class));

        clientsDS.show();
        clientsDS.printSchema();

        //   Filtrer les clients adultes (typage fort)
        Dataset<Client> adultsDS = clientsDS.filter((FilterFunction<Client>) client -> client.getAge() >= 18);
        adultsDS.show();


        Dataset<Client> adultsDS2 = clientsDS.filter(Client::isLegalAge);
        adultsDS2.show();

        // Afficher les clients adultes
        adultsDS.show();

        // Fermer la SparkSession
        spark.stop();
    }
}